File: javascript-lib/src/ws.js
// Server-side WebSocket (RFC 6455) byte-stream adapter for aiomsg bind sockets
// (see ../PROTOCOL.md §10).
//
// sniffAndUpgrade() peeks the first byte of an accepted connection and decides
// whether the peer speaks raw aiomsg or HTTP. On HTTP it performs the WebSocket
// upgrade and hands back a WsDuplex — a Duplex stream that presents the *same*
// interface the rest of the socket uses (write() / 'data' / 'close' / destroy),
// so the connection handler, codec, and every higher layer are untouched.
//
// WebSocket message boundaries carry no meaning: inbound binary-message payloads
// are concatenated and emitted as one byte stream (the stream of §2), and each
// aiomsg write goes out as one unmasked binary frame. Only the server half of
// RFC 6455 is implemented; no subprotocol or extension is ever negotiated.

import { createHash } from "node:crypto";
import { Duplex } from "node:stream";

const WS_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11";

// Opcodes (RFC 6455 §5.2).
const OP_CONT = 0x0;
const OP_TEXT = 0x1;
const OP_BIN = 0x2;
const OP_CLOSE = 0x8;
const OP_PING = 0x9;
const OP_PONG = 0xa;

const MAX_REQUEST_BYTES = 8192; // cap the pre-upgrade HTTP request

// --- Handshake (pure) --------------------------------------------------------

// base64(SHA1(key + GUID)) per RFC 6455 §4.2.2.
export function computeAccept(key) {
  return createHash("sha1").update(key + WS_GUID).digest("base64");
}

// Validate an HTTP upgrade request (a Buffer/string through the blank line) and
// return its Sec-WebSocket-Key. Throws Error whose message is the HTTP status
// line to send on any violation. Host/Origin/Sec-WebSocket-Protocol/-Extensions
// are ignored: no subprotocol or extension is negotiated, and auth is out of
// scope (WEBSOCKET-PLAN.md §1.5).
export function parseUpgrade(request) {
  const text = Buffer.isBuffer(request) ? request.toString("latin1") : String(request);
  const lines = text.split("\r\n");
  const parts = lines[0].split(" ");
  if (parts.length !== 3 || parts[0] !== "GET" || !parts[2].startsWith("HTTP/1.")) {
    throw new Error("400 Bad Request");
  }
  const headers = new Map();
  for (const line of lines.slice(1)) {
    if (line === "") continue;
    const i = line.indexOf(":");
    if (i === -1) throw new Error("400 Bad Request");
    headers.set(line.slice(0, i).trim().toLowerCase(), line.slice(i + 1).trim());
  }
  if ((headers.get("upgrade") || "").toLowerCase() !== "websocket") {
    throw new Error("400 Bad Request");
  }
  const connectionTokens = (headers.get("connection") || "")
    .split(",")
    .map((t) => t.trim().toLowerCase());
  if (!connectionTokens.includes("upgrade")) throw new Error("400 Bad Request");
  if ((headers.get("sec-websocket-version") || "") !== "13") {
    throw new Error("426 Upgrade Required");
  }
  const key = headers.get("sec-websocket-key") || "";
  if (Buffer.from(key, "base64").length !== 16) throw new Error("400 Bad Request");
  return key;
}

export function successResponse(key) {
  return Buffer.from(
    "HTTP/1.1 101 Switching Protocols\r\n" +
      "Upgrade: websocket\r\n" +
      "Connection: Upgrade\r\n" +
      `Sec-WebSocket-Accept: ${computeAccept(key)}\r\n` +
      "\r\n",
    "latin1",
  );
}

export function errorResponse(status) {
  const body = status.startsWith("426")
    ? `HTTP/1.1 ${status}\r\nSec-WebSocket-Version: 13\r\n\r\n`
    : `HTTP/1.1 ${status}\r\n\r\n`;
  return Buffer.from(body, "latin1");
}

// --- Frame layer (pure) ------------------------------------------------------

// Encode one server->client frame: FIN=1, RSV=0, unmasked, given opcode.
export function encodeServerFrame(opcode, payload) {
  const n = payload.length;
  let header;
  // TODO(frame-size): enforce a configurable maximum frame length
  if (n < 126) {
    header = Buffer.from([0x80 | opcode, n]);
  } else if (n < 65536) {
    header = Buffer.from([0x80 | opcode, 126, (n >> 8) & 0xff, n & 0xff]);
  } else {
    header = Buffer.alloc(10);
    header[0] = 0x80 | opcode;
    header[1] = 127;
    header.writeBigUInt64BE(BigInt(n), 2);
  }
  return Buffer.concat([header, payload]);
}

// --- Adapter -----------------------------------------------------------------

class WsDuplex extends Duplex {
  constructor(raw) {
    super();
    this._raw = raw;
    this._inbuf = Buffer.alloc(0);
    this._closeSent = false;

    raw.on("data", (chunk) => this._onRaw(chunk));
    raw.once("error", (err) => this.destroy(err));
    // Peer TCP close: end the readable side and emit our own 'close' (the
    // connection handler tears down on it).
    raw.once("close", () => this.destroy());
  }

  setNoDelay(v) {
    this._raw.setNoDelay(v);
    return this;
  }

  _read() {
    if (this._raw.isPaused()) this._raw.resume();
  }

  // Each aiomsg write (already LENGTH || ENVELOPE bytes) becomes one binary
  // frame; there is no fragmentation on the sending side (§2.2).
  _write(chunk, _enc, cb) {
    const ok = this._raw.write(encodeServerFrame(OP_BIN, chunk));
    if (ok) cb();
    else this._raw.once("drain", cb);
  }

  _destroy(err, cb) {
    this._sendClose(1000);
    this._raw.destroy();
    cb(err);
  }

  _sendClose(code) {
    if (this._closeSent) return;
    this._closeSent = true;
    const payload = code == null ? Buffer.alloc(0) : Buffer.from([(code >> 8) & 0xff, code & 0xff]);
    try {
      this._raw.write(encodeServerFrame(OP_CLOSE, payload));
    } catch {
      /* peer already gone */
    }
  }

  _fail(code) {
    this._sendClose(code);
    this.destroy(); // _destroy tears down the raw socket and emits 'close'
  }

  _onRaw(chunk) {
    this._inbuf = this._inbuf.length ? Buffer.concat([this._inbuf, chunk]) : chunk;
    for (;;) {
      const frame = this._parseFrame();
      if (frame === null) break; // need more bytes
      if (frame === false) return; // protocol error already handled
      const { opcode, payload } = frame;
      if (opcode === OP_BIN || opcode === OP_CONT) {
        if (!this.push(payload)) this._raw.pause();
      } else if (opcode === OP_PING) {
        this._raw.write(encodeServerFrame(OP_PONG, payload));
      } else if (opcode === OP_PONG) {
        /* ignore */
      } else if (opcode === OP_TEXT) {
        this._fail(1003); // text is not valid for aiomsg
        return;
      } else if (opcode === OP_CLOSE) {
        const code = payload.length >= 2 ? payload.readUInt16BE(0) : 1000;
        this._sendClose(code);
        this.destroy();
        return;
      } else {
        this._fail(1002); // unknown opcode
        return;
      }
    }
  }

  // Try to parse one complete client frame from _inbuf. Returns the frame, null
  // if more bytes are needed, or false if a protocol error was raised.
  _parseFrame() {
    const buf = this._inbuf;
    if (buf.length < 2) return null;
    const b0 = buf[0];
    const b1 = buf[1];
    if (b0 & 0x70) return this._fail(1002), false; // RSV bits set
    const opcode = b0 & 0x0f;
    const fin = (b0 & 0x80) !== 0;
    if (!(b1 & 0x80)) return this._fail(1002), false; // client frames MUST be masked
    let len = b1 & 0x7f;
    let offset = 2;
    if (len === 126) {
      if (buf.length < 4) return null;
      len = buf.readUInt16BE(2);
      offset = 4;
    } else if (len === 127) {
      if (buf.length < 10) return null;
      if (buf[2] & 0x80) return this._fail(1002), false; // 64-bit length MSB must be 0
      // TODO(frame-size): enforce a configurable maximum frame length
      len = Number(buf.readBigUInt64BE(2));
      offset = 10;
    }
    if (opcode >= 0x8 && (!fin || len > 125)) return this._fail(1002), false; // control frame limits
    if (buf.length < offset + 4 + len) return null; // mask (4) + payload
    const mask = buf.subarray(offset, offset + 4);
    const masked = buf.subarray(offset + 4, offset + 4 + len);
    const payload = Buffer.allocUnsafe(len);
    for (let i = 0; i < len; i++) payload[i] = masked[i] ^ mask[i & 3];
    this._inbuf = buf.subarray(offset + 4 + len);
    return { opcode, fin, payload };
  }
}

// --- Entry point -------------------------------------------------------------

// Sniff an accepted socket and resolve to { kind, socket } where kind is "raw"
// (socket is the original net/tls socket, first byte pushed back) or "ws"
// (socket is a WsDuplex), or null if the connection should be dropped
// (unknown first byte, EOF, or a rejected upgrade). PROTOCOL.md §10.
export function sniffAndUpgrade(raw) {
  return new Promise((resolve) => {
    let buf = Buffer.alloc(0);
    let settled = false;
    const cleanup = () => {
      raw.removeListener("readable", onReadable);
      raw.removeListener("error", onDone);
      raw.removeListener("end", onDone);
      raw.removeListener("close", onDone);
    };
    const finish = (value) => {
      if (settled) return;
      settled = true;
      cleanup();
      resolve(value);
    };
    const onDone = () => finish(null);

    // Read in PAUSED mode ('readable' + read()) so the socket never enters
    // flowing mode: bytes we don't consume stay buffered for the handler that
    // takes over, with none dropped in the handoff window.
    const onReadable = () => {
      let chunk;
      while ((chunk = raw.read()) !== null) buf = buf.length ? Buffer.concat([buf, chunk]) : chunk;
      if (buf.length === 0) return;
      const first = buf[0];

      if (first === 0x00) {
        // Raw aiomsg: hand the already-read bytes back as `prebuffer`; the
        // socket stays paused until the handler attaches its own 'data' reader.
        finish({ kind: "raw", socket: raw, prebuffer: buf });
        return;
      }
      if (first !== 0x47 /* 'G' */) {
        raw.destroy();
        finish(null);
        return;
      }

      const idx = buf.indexOf("\r\n\r\n");
      if (idx === -1) {
        if (buf.length > MAX_REQUEST_BYTES) {
          raw.write(errorResponse("400 Bad Request"));
          raw.destroy();
          finish(null);
        }
        return; // await the rest of the request
      }
      const request = buf.subarray(0, idx + 4);
      const leftover = buf.subarray(idx + 4);
      let key;
      try {
        key = parseUpgrade(request);
      } catch (e) {
        raw.write(errorResponse(e.message));
        raw.destroy();
        finish(null);
        return;
      }
      cleanup(); // stop sniffing before the adapter takes over the raw stream
      raw.write(successResponse(key));
      const adapter = new WsDuplex(raw);
      resolve({ kind: "ws", socket: adapter });
      // Feed the bytes read past the request (the first WS frames) to the
      // adapter's parser, but only after the connection handler has attached
      // its listeners. setImmediate runs after all await continuations, so a
      // leftover close/error frame still reaches an attached 'close' handler.
      if (leftover.length) setImmediate(() => adapter._onRaw(leftover));
    };

    raw.on("readable", onReadable);
    raw.once("error", onDone);
    raw.once("end", onDone);
    raw.once("close", onDone);
  });
}

File: golang-lib/ws.go
package aiomsg

// Server-side WebSocket (RFC 6455) byte-stream adapter for aiomsg bind sockets
// (see PROTOCOL.md §10).
//
// sniffAndWrap peeks the first byte of an accepted connection and decides
// whether the peer speaks raw aiomsg or HTTP. On HTTP it performs the WebSocket
// upgrade and returns a net.Conn whose Read/Write carry the aiomsg byte stream
// inside binary WebSocket frames, so handleConnection, the codec, and every
// higher layer are untouched — the adapter is a pure transport shim, exactly
// like the TLS layer.
//
// WebSocket message boundaries carry no meaning: inbound binary-message payloads
// are concatenated into one byte stream (the stream of PROTOCOL.md §2), and each
// aiomsg frame (one writeFrame call) goes out as one unmasked binary frame. Only
// the server half of RFC 6455 is implemented; no subprotocol or extension is
// ever negotiated. This file is hand-rolled and depends only on the standard
// library.

import (
	"bufio"
	"crypto/sha1"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"io"
	"net"
	"strings"
	"sync"
	"time"
)

// wsGUID is appended to the client key before SHA-1 (RFC 6455 §1.3).
const wsGUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

// Opcodes (RFC 6455 §5.2).
const (
	opCont  = 0x0
	opText  = 0x1
	opBin   = 0x2
	opClose = 0x8
	opPing  = 0x9
	opPong  = 0xA
)

const maxRequestBytes = 8192 // cap on the pre-upgrade HTTP request

// errUnknownProtocol is returned when the first byte is neither raw aiomsg
// (0x00) nor the start of an HTTP request ('G').
var errUnknownProtocol = errors.New("aiomsg: unrecognised protocol on accept")

// wsError carries the HTTP status line to return for a rejected upgrade.
type wsError struct{ status string }

func (e wsError) Error() string { return "aiomsg: bad ws upgrade: " + e.status }

// --- Handshake (pure) -------------------------------------------------------

// computeAccept is base64(SHA1(key + GUID)) per RFC 6455 §4.2.2. Test vector:
// key "dGhlIHNhbXBsZSBub25jZQ==" -> "s3pPLMBiTxaQ9kYGzzhZRbK+xOo=".
func computeAccept(key string) string {
	sum := sha1.Sum([]byte(key + wsGUID))
	return base64.StdEncoding.EncodeToString(sum[:])
}

// readUpgrade reads and validates the HTTP upgrade request from br, returning
// the Sec-WebSocket-Key. Any bytes past the blank line stay buffered in br (they
// are the first WS frames). On any violation it returns a wsError whose status
// is the HTTP status line to send. Host/Origin/Sec-WebSocket-Protocol and
// -Extensions are ignored: no subprotocol or extension is negotiated, and auth
// is out of scope (WEBSOCKET-PLAN.md §1.5).
func readUpgrade(br *bufio.Reader) (string, error) {
	var lines []string
	total := 0
	for {
		line, err := br.ReadString('\n')
		if err != nil {
			return "", wsError{"400 Bad Request"}
		}
		total += len(line)
		if total > maxRequestBytes {
			return "", wsError{"400 Bad Request"}
		}
		line = strings.TrimRight(line, "\r\n")
		if line == "" {
			break // end of header block
		}
		lines = append(lines, line)
	}

	parts := strings.Fields(lines[0])
	if len(parts) != 3 || parts[0] != "GET" || !strings.HasPrefix(parts[2], "HTTP/1.") {
		return "", wsError{"400 Bad Request"}
	}
	headers := make(map[string]string)
	for _, l := range lines[1:] {
		i := strings.IndexByte(l, ':')
		if i < 0 {
			return "", wsError{"400 Bad Request"}
		}
		headers[strings.ToLower(strings.TrimSpace(l[:i]))] = strings.TrimSpace(l[i+1:])
	}

	if !strings.EqualFold(headers["upgrade"], "websocket") {
		return "", wsError{"400 Bad Request"}
	}
	if !hasToken(headers["connection"], "upgrade") {
		return "", wsError{"400 Bad Request"}
	}
	if headers["sec-websocket-version"] != "13" {
		return "", wsError{"426 Upgrade Required"}
	}
	key := headers["sec-websocket-key"]
	if decoded, err := base64.StdEncoding.DecodeString(key); err != nil || len(decoded) != 16 {
		return "", wsError{"400 Bad Request"}
	}
	return key, nil
}

// hasToken reports whether a comma-separated header value contains token
// (case-insensitive), e.g. Connection: keep-alive, Upgrade.
func hasToken(value, token string) bool {
	for _, t := range strings.Split(value, ",") {
		if strings.EqualFold(strings.TrimSpace(t), token) {
			return true
		}
	}
	return false
}

func successResponse(key string) []byte {
	return []byte("HTTP/1.1 101 Switching Protocols\r\n" +
		"Upgrade: websocket\r\n" +
		"Connection: Upgrade\r\n" +
		"Sec-WebSocket-Accept: " + computeAccept(key) + "\r\n" +
		"\r\n")
}

func errorResponse(status string) []byte {
	if strings.HasPrefix(status, "426") {
		return []byte("HTTP/1.1 " + status + "\r\nSec-WebSocket-Version: 13\r\n\r\n")
	}
	return []byte("HTTP/1.1 " + status + "\r\n\r\n")
}

// --- Frame layer (pure) -----------------------------------------------------

// encodeServerFrame builds one server->client frame: FIN=1, RSV=0, unmasked.
func encodeServerFrame(opcode byte, payload []byte) []byte {
	n := len(payload)
	var header []byte
	// TODO(frame-size): enforce a configurable maximum frame length
	switch {
	case n < 126:
		header = []byte{0x80 | opcode, byte(n)}
	case n < 65536:
		header = []byte{0x80 | opcode, 126, byte(n >> 8), byte(n)}
	default:
		header = make([]byte, 10)
		header[0] = 0x80 | opcode
		header[1] = 127
		binary.BigEndian.PutUint64(header[2:], uint64(n))
	}
	return append(header, payload...)
}

func closePayload(code int) []byte {
	if code == 0 {
		return nil
	}
	return []byte{byte(code >> 8), byte(code)}
}

// --- Adapters ---------------------------------------------------------------

// rawConn presents an accepted raw-aiomsg connection through the bufio.Reader
// that already holds the sniffed first byte, so the codec sees the whole stream
// (net.Conn has no "unread"). All other methods delegate to the embedded conn.
type rawConn struct {
	net.Conn
	br *bufio.Reader
}

func (c *rawConn) Read(p []byte) (int, error) { return c.br.Read(p) }

// wsConn is a net.Conn whose Read yields the concatenation of inbound binary
// WebSocket payloads and whose Write emits one binary frame per call. Control
// frames (ping/pong/close) are handled transparently inside Read. Reads run on
// the connection's single reader goroutine; a mutex serialises the frames the
// reader (pong/close) and the writer goroutine put on the wire.
type wsConn struct {
	net.Conn
	br        *bufio.Reader
	inbuf     []byte
	writeMu   sync.Mutex // serialises frames written by the reader and writer
	closeOnce sync.Once  // guards a single outbound close frame
}

func newWSConn(inner net.Conn, br *bufio.Reader) *wsConn {
	return &wsConn{Conn: inner, br: br}
}

func (c *wsConn) Read(p []byte) (int, error) {
	for len(c.inbuf) == 0 {
		if err := c.readNextFrame(); err != nil {
			return 0, err
		}
	}
	n := copy(p, c.inbuf)
	c.inbuf = c.inbuf[n:]
	return n, nil
}

// readNextFrame reads one client frame and either appends its payload to inbuf
// or handles it (ping->pong, pong ignored, close/text/protocol-error -> EOF).
func (c *wsConn) readNextFrame() error {
	var h [2]byte
	if _, err := io.ReadFull(c.br, h[:]); err != nil {
		return err
	}
	b0, b1 := h[0], h[1]
	if b0&0x70 != 0 { // any RSV bit set
		return c.fail(1002)
	}
	opcode := b0 & 0x0f
	fin := b0&0x80 != 0
	if b1&0x80 == 0 { // client frames MUST be masked
		return c.fail(1002)
	}
	length := int(b1 & 0x7f)
	switch length {
	case 126:
		var ext [2]byte
		if _, err := io.ReadFull(c.br, ext[:]); err != nil {
			return err
		}
		length = int(binary.BigEndian.Uint16(ext[:]))
	case 127:
		var ext [8]byte
		if _, err := io.ReadFull(c.br, ext[:]); err != nil {
			return err
		}
		if ext[0]&0x80 != 0 { // 64-bit length MSB must be 0
			return c.fail(1002)
		}
		// TODO(frame-size): enforce a configurable maximum frame length
		length = int(binary.BigEndian.Uint64(ext[:]))
	}
	if opcode >= 0x8 && (!fin || length > 125) { // control-frame limits
		return c.fail(1002)
	}

	var mask [4]byte
	if _, err := io.ReadFull(c.br, mask[:]); err != nil {
		return err
	}
	payload := make([]byte, length)
	if _, err := io.ReadFull(c.br, payload); err != nil {
		return err
	}
	for i := range payload {
		payload[i] ^= mask[i&3]
	}

	switch opcode {
	case opBin, opCont:
		c.inbuf = append(c.inbuf, payload...)
		return nil
	case opPing:
		c.writeControl(opPong, payload)
		return nil
	case opPong:
		return nil
	case opText:
		return c.fail(1003) // text is not valid for aiomsg
	case opClose:
		code := 1000
		if len(payload) >= 2 {
			code = int(binary.BigEndian.Uint16(payload))
		}
		c.sendClose(code)
		return io.EOF
	default:
		return c.fail(1002) // unknown opcode
	}
}

func (c *wsConn) fail(code int) error {
	c.sendClose(code)
	return io.EOF
}

func (c *wsConn) Write(p []byte) (int, error) {
	frame := encodeServerFrame(opBin, p)
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	if _, err := c.Conn.Write(frame); err != nil {
		return 0, err
	}
	return len(p), nil
}

func (c *wsConn) writeControl(opcode byte, payload []byte) {
	frame := encodeServerFrame(opcode, payload)
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	_, _ = c.Conn.Write(frame)
}

// sendClose writes at most one close frame for the life of the connection.
func (c *wsConn) sendClose(code int) {
	c.closeOnce.Do(func() { c.writeControl(opClose, closePayload(code)) })
}

func (c *wsConn) Close() error {
	c.sendClose(1000)
	return c.Conn.Close()
}

// --- Entry point ------------------------------------------------------------

// sniffAndWrap detects raw-vs-WebSocket on an accepted connection (PROTOCOL.md
// §10) and returns the net.Conn the connection handler should drive: the raw
// connection (first byte pushed back) or a wsConn. It returns an error if the
// connection should be closed (unknown first byte, EOF, timeout, or a rejected
// upgrade — for which the HTTP error response is written first).
func sniffAndWrap(conn net.Conn) (net.Conn, error) {
	// Bound the whole sniff + upgrade so a silent client cannot hold the slot.
	_ = conn.SetReadDeadline(time.Now().Add(heartbeatTimeout))
	br := bufio.NewReaderSize(conn, maxRequestBytes)
	first, err := br.Peek(1)
	if err != nil {
		return nil, err
	}

	switch first[0] {
	case 0x00:
		setNoDelay(conn)
		_ = conn.SetReadDeadline(time.Time{}) // handleConnection sets its own
		return &rawConn{Conn: conn, br: br}, nil
	case 'G':
		key, err := readUpgrade(br)
		if err != nil {
			if we, ok := err.(wsError); ok {
				_, _ = conn.Write(errorResponse(we.status))
			}
			return nil, err
		}
		if _, err := conn.Write(successResponse(key)); err != nil {
			return nil, err
		}
		setNoDelay(conn)
		_ = conn.SetReadDeadline(time.Time{})
		return newWSConn(conn, br), nil
	default:
		return nil, errUnknownProtocol
	}
}
